### PR TITLE
added the cord to fit wavelength solution for wavelength calibration

### PR DESCRIPTION
**Pull request by YuiKasagi**
I added the cord to convert pixels to wavelength by fitting polynomial function (Legendre).
I would like to describe it in #11

### CODE REVIEW

File: src/pyird/image/wavcal.py [hunk 1]
```diff
@@ -1,11 +1,85 @@
...
 from astropy.io import fits
 import numpy as np
 from scipy.signal import medfilt
+from scipy.optimize import leastsq
 import pandas as pd
 
+import matplotlib.pyplot as plt
+
+def pdat2matrix(pdat):
+    data = np.zeros((l-j,2048))
```
Comment (HajimeKawahara) [inline src/pyird/image/wavcal.py:11]: can you use a parameter for 2048 to be able to change the detector?
like
```python
pdat2matrix(pdat, npix=2048)
    data = np.zeros((l-j,npix))
```

File: src/pyird/image/wavcal.py [hunk 2]
```diff
@@ -1,11 +1,85 @@
+from pyird.io.read_linelist import read_linelist
 from astropy.io import fits
 import numpy as np
 from scipy.signal import medfilt
+from scipy.optimize import leastsq
 import pandas as pd
 
+import matplotlib.pyplot as plt
+
+def pdat2matrix(pdat):
```
Comment (HajimeKawahara) [inline src/pyird/image/wavcal.py:10]: Can you leave the google/napoleon type comment for all of the functions?
like [this one](https://github.com/prvjapan/pyird/blob/cc9dbbeeda312e67e0f6f76f33f520fbb042b97e/src/pyird/spec/rsdmat.py#L6)

These comments enable us to generate the API document automatically, in `documents`.

Comment (HajimeKawahara) [reply src/pyird/image/wavcal.py:10]: The API example is [here](http://secondearths.sakura.ne.jp/pyird/pyird/pyird.html).

File: src/pyird/image/wavcal.py [hunk 3]
```diff
@@ -1,11 +1,85 @@
...
+        order, pixel, wav = pdat.iloc[i].values
+        order = order - m0
+        data[order,pixel] = wav
+    return data
+
+def fitfunc(XY,Ni,Nx,*params):
+    pix = XY[0][0]
+    m = XY[1][:,0]
+    minv = 1/XY[1]
+    p=[]
```
Comment (HajimeKawahara) [inline src/pyird/image/wavcal.py:23]:
```python
p = np.array(params).reshape((Ni,Nx))
```
is enough?

File: src/pyird/image/wavcal.py [hunk 4]
```diff
@@ -1,11 +1,85 @@
...
+
+def errfunc(p,XY,data,Ni,Nx):
+    ind_data = data.ravel()!=0
+    data_on = data.ravel()[ind_data]
+    model = fitfunc(XY,Ni,Nx,p)
+    model_on = model[ind_data]
+    return data_on - model_on
+
+def fit_wav_solution(XY,data,Ni,Nx):
+    p0 = np.array([1]*Ni*Nx)
```
Comment (HajimeKawahara) [inline src/pyird/image/wavcal.py:38]:
```python
p0 = np.ones(Ni*Nx)
````

File: src/pyird/image/wavcal.py [hunk 5]
```diff
@@ -1,11 +1,85 @@
...
+        if data_tmp != 0:
+            res = data_tmp-wavsol[i]
+            if np.abs(res)>Nsigma:
+                #print(n, data_tmp)
+                drop_ind.append(n)
+            n += 1
+    return residuals,drop_ind
+
+def plot_result(wavsol,data):
+    wavsol_2d = wavsol.reshape(l-j,2048)
```
Comment (HajimeKawahara) [inline src/pyird/image/wavcal.py:59]: define a new variable for 2048

File: src/pyird/image/wavcal.py [hunk 6]
```diff
@@ -1,11 +1,85 @@
...
+            res = data_tmp-wavsol[i]
+            ax2.plot(pix,res,".",color="tab:blue")
+    ax2.set(xlabel="pixel",ylabel="residual [nm]")
+    plt.savefig('wavcal_final.png')
+    plt.show()
+
 if __name__ == '__main__':
     # Load a Th-Ar spectrum
-    tharfile = '/Users/kasagiyui/IRD/PhDwork/pyird/data/20210317/thar_mmf2_a_h.fits'
+    tharfile = '/Users/yuikasagi/IRD/PhDwork/pyird/data/20210317/thar_mmf2_a_h.fits'
```
Comment (HajimeKawahara) [inline src/pyird/image/wavcal.py:82]: Can you put this file somewhere?
I wanna test the PR.

File: src/pyird/image/wavcal.py (hunk 1 above)
Comment (YuiKasagi) [reply src/pyird/image/wavcal.py]: I adopted npix. In addition, I modified the shape of matrix (order, pixel) -> (pixel, order).

File: src/pyird/image/wavcal.py (hunk 2 above)
Comment (YuiKasagi) [reply src/pyird/image/wavcal.py]: added the comment for all of the functions. please check it out.

File: src/pyird/image/wavcal.py (hunk 3 above)
Comment (YuiKasagi) [reply src/pyird/image/wavcal.py]: Thank you for pointed it out. I have changed it.

File: src/pyird/image/wavcal.py (hunk 4 above)
Comment (YuiKasagi) [reply src/pyird/image/wavcal.py]: revised

File: src/pyird/image/wavcal.py (hunk 5 above)
Comment (YuiKasagi) [reply src/pyird/image/wavcal.py]: revised

File: src/pyird/image/wavcal.py (hunk 6 above)
Comment (YuiKasagi) [reply src/pyird/image/wavcal.py]: I put it in 'pyird/data/thar_mmf2_a_H_20210317.fits'. Maybe it will work.

Comment (HajimeKawahara) [reply src/pyird/image/wavcal.py]: I think this file should be external because of its size.

Comment (HajimeKawahara) [reply src/pyird/image/wavcal.py]: Can you add the file name in MANIFEST.in like [this](https://github.com/prvjapan/pyird/blob/master/MANIFEST.in).
Then, you can call the file path by 
```python
import pkg_resources
path = (pkg_resources.resource_filename('pyird', 'data/thar_mmf2_a_H_20210317.fits'))
```

File: src/pyird/image/wavcal.py (hunk 2 above)
Comment (HajimeKawahara) [reply src/pyird/image/wavcal.py]: Thanks. 
```python
Return:
```
should be 
```python
Returns:
```
probably....

**Review by HajimeKawahara — APPROVED** [review]

**Pull request merged**